src/module_wrap.cc

```diff
@@ -705,15 +903,42 @@ void ModuleWrap::Resolve(const FunctionCallbackInfo<Value>& args) {
         env, "second argument is not a URL string");
   }
 
-  Maybe<URL> result = node::loader::Resolve(env, specifier_std, url);
-  if (result.IsNothing() || (result.FromJust().flags() & URL_FLAGS_FAILED)) {
-    std::string msg = "Cannot find module " + specifier_std;
-    return node::THROW_ERR_MISSING_MODULE(env, msg.c_str());
+  TryCatchScope try_catch(env);
+  Maybe<URL> result =
+      node::loader::Resolve(env,
+                            specifier_std,
+                            url);
+  if (result.IsNothing()) {
+    CHECK(try_catch.HasCaught());
+    try_catch.ReThrow();
```

Comment:
Why is this `try_catch` is being added if we’re going to re-throw the exception either way?

If it is necessary: It can be a bit brittle to rely on the fact that `result.IsNothing()` implies `try_catch.HasCaught()` – for V8 APIs, `result.IsNothing()` implies that an exception is pending, but it does *not* imply that that exception wasn’t already pending before the `TryCatch`, meaning that `try_catch.HasCaught()` might be false. (This has caused trouble with `worker.terminate()`, see e.g. 7b198935d63aec4acace1500d81ac2662c732d18.)

Reply:
The `IsNothing()` here is actually entirely about our own API contracts, and nothing to do with v8. And the TryCatch is there to ensure we continue to abide by that contract, to stop this drifting in future. That is, all module_wrap resolver errors will both throw an error and return Nothing(), and never just one or the other.

Reply:
@guybedford Okay, I understand – unfortunately, that doesn’t really negate that this code can hard-crash the process. :/

You can add `env->isolate()->TerminateExecution();` before `TryCatchScope try_catch(env);` and run the tests – execution termination can happen at any point with `worker.terminate()`, so this is be a valid scenario to consider.

I originally did that to try and trigger the `CHECK(try_catch.HasCaught())` here, but there seem to be more issues, but apparently you don’t do any V8 calls that would fail under that circumstance (anything that would actually call into JS)? It *does* crash when calling `resolution.ToObject(env).ToLocalChecked()` below, though.

I’m also worried about accidentally overriding the termination exception using `.ReThrow()`. That can make it look as if `throw null;` was what happened, when really it should just stop the current thread.

Reply:
> Okay, I understand – unfortunately, that doesn’t really negate that this code can hard-crash the process. :/

I'm still not sure I understand the failure case that you are concerned about here. Could you clarify how that might happen here? Do you mean if something else caused an error and we didn't get a Nothing return? Although even in that case, surely the error would propagate fine on returning to JS?

> I’m also worried about accidentally overriding the termination exception using .ReThrow()

Do you mean that if something else abruptly terminated Node, and then our ReThrow might interfere with that? Would simply removing the TryCatch / ReThrow mitigate that?

Reply:
@addaleax would overriding ReThrow to check for IsTerminated in our TryCatchScope work? given the current usage of ReThrow in our codebase i feel like it could be a nice refactoring in either case if it makes sense.

Reply:
> > Okay, I understand – unfortunately, that doesn’t really negate that this code can hard-crash the process. :/
> 
> I'm still not sure I understand the failure case that you are concerned about here. Could you clarify how that might happen here? Do you mean if something else caused an error and we didn't get a Nothing return? Although even in that case, surely the error would propagate fine on returning to JS?

What might happen here is that a termination exception may be raised, possibly before the `TryCatchScope`. The ways to cause that would be another thread calling `worker.terminate()` or `Stop()` from the embedder API.

If an exception, include a possible termination exception, is pending, V8 calls that would enter JS will return empty `Maybe`s/`MaybeLocal`s, because there already *is* an exception. However, the `TryCatch` would not receive the exception, because nothing happened inside this scope – and so the `CHECK()` for that would fail.

> 
> > I’m also worried about accidentally overriding the termination exception using .ReThrow()
> 
> Do you mean that if something else abruptly terminated Node, and then our ReThrow might interfere with that? Would simply removing the TryCatch / ReThrow mitigate that?

Yes, exactly. I think what would happen is that the `ReThrow()` would be equivalent to `throw null;`, which overrides the pending termination exception. Removing it would mitigate that.

Reply:
@devsnek Yeah, I think that would make sense. I’m also not fully sure if this is intended on the V8 side to begin with.

Reply:
Thanks for explaining, that makes sense.

Since we aren't crossing into JS though, surely this can never happen? Or is the risk that we might inadvertently add such a code path in future?

We could simply remove the TryCatch scope entirely certainly, and then just effectively have the implicit undefined return on the Nothing case, which would be caught as a bug in the JS side anyway if that ever happened in future. So I don't see any risk in removing it entirely.

Reply:
> Since we aren't crossing into JS though, surely this can never happen? Or is the risk that we might inadvertently add such a code path in future?

This can hide pretty well, or change subtly – for example, the `resolution.ToObject()` call below calls into JS, but that isn’t immediately visible (or could change at any point), or a `Get()` or `Set()` call could end up calling JS getters/setters.

Reply:
(So, yes, my preference would be to remove the `TryCatchScope` here.)

Reply:
This is fixed afaict. PTAL and resolve convo if both of you are happy with f180ddc